[Merged by Bors] - chore(scripts): require all scripts to be documented in the README

grunweg opened the pull request:
This is valuable so
- contributors can discover existing useful tools,
- existing knowledge is documented and not lost
- it (hopefully) inspires to re-consider if all each script is useful/why we have it. Having a [bounded number of entry-points](https://matklad.github.io/2023/12/31/O(1)-build-file.html) to CI is useful to have.

This is implemented as a new check, piggy-backing the `lint_style` executable.

To make CI pass, this PR adds a README file to `scripts`, containing stub descriptions for the current files. Future PRs should fill this out.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks, and is labeled with `awaiting-review`.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg commented:
@semorrison You may the interested in this one: obviously, it's not ready to land yet.

kim-em commented:
Apologies this rotted. I just merged `master`, hopefully it works.

I like this! Please merge if you are too, otherwise we can discuss what remains to be done.

bors d+

grunweg commented:
@kim-em Thanks; I just merged master and fixed up this PR. (In merging master, you made the check never run - I assume that was not what you intended, right?) After actually running, the check rightfully errors about undocumented scripts. Thus, I added a stub README with stub entries for each script (clustering them a bit, but not writing documentation).

I presume this matches your intention. Can you confirm, please? (And would you like to find a volunteer to document at least the scripts for adaption PRs; I don't know these at all.)

grunweg commented:
This PR adds a stub README to the `scripts` directory, with a big TODO to document the various scripts. It would be good to document what these scripts actually do (regardless of whether documenting every file is enforced). I need some help for this:
- @kim-em or @jcommelin Do you mind writing two lines about `create-adaptation-pr.sh`, `lean-pr-testing-comments.sh`? Either commenting on this PR (or adding a commit directly), or PRing them as a follow-up is fine.
- @adomani I think you wrote a fair number of the cluster `declarations_diff.sh`, `count-trans-deps.py`, `import_trans_difference.sh`, `update_PR_comment.sh` and `import-graph-report.py`. I know that they are used to update the sticky PR comment, but not how these scripts call each other. Do you mind adding a few words about this?
- I'm happy to take `lint-style.py`, `lint-style.lean` and `print-style-errors.sh`.
- Contributions documenting other scripts are certainly welcome, from everybody.

Feel free to look at the stub README and document a few more, if you like. (I may do so myself, but not this week.)
Thanks for helping out!

jcommelin commented on scripts/README.md:
```diff
@@ -0,0 +1,39 @@
+- `technical-debt-metrics.sh`
+- `init_creation.sh` (supposed to be run manually; might be removed soon?)
+
+- `create-adaptation-pr.sh`, `lean-pr-testing-comments.sh`
```
`create-adaptation-pr.sh` implements some of the steps in the workflow described at https://leanprover-community.github.io/contribute/tags_and_branches.html#mathlib-nightly-and-bump-branches
Specifically, it will:
- merge `master` into `bump/v4.x.y`
- create a new branch off of `bump/v4.x.y`
- merge `nightly-testing` into the new branch
- open a PR to merge the new branch back into `bump/v4.x.y`
- announce the PR on zulip

If there are merge conflicts, it pauses and asks for help from the human driver.

kim-em replied:
I've moved this into the README

grunweg commented:
Thank you for documenting most entries! I just took a pass over the ones I know, and formatted the file consistently. This is just missing an entry on ~four scripts used to generate the PR summary comment. Help on these is still welcome - though in my opinion, this is already useful.

From my point of view, this can be merged. @bryangingechen I saw you self-assigned this. If I don't hear objections from you, I will merge this in a day (or so).

adomani commented:
I pushed a possible categorization of the remaining scripts with some details.  Let me know if this is good!  A lot of these files also have individual docs, so the combination of the generic dispatch and the specific information is probably good enough.

grunweg commented on scripts/README.md:
```diff
@@ -0,0 +1,87 @@
+  It is used by the `PR_summary` workflow to maintain an up-to-date report with a searchble history.
+- `get_tlabel.sh` extracts the `t-`label that a PR has (assuming that there is exactly one).
+  It is used by the `maintainer_merge` family of workflows to dispatch `maintainer merge` requests
+  to the appropriate topic.
```
```suggestion
  to the appropriate topic on zulip.

grunweg commented on scripts/README.md:
```diff
@@ -0,0 +1,87 @@
+
+**API surrounding CI**
+- `update_PR_comment.sh` is a script that edits an existing message (or creates a new one).
+  It is used by the `PR_summary` workflow to maintain an up-to-date report with a searchble history.
```
```suggestion
  It is used by the `PR_summary` workflow to maintain an up-to-date report with a searchable history.

grunweg commented:
Thanks for adding the last missing docs! I like the changes and just pushed a last tweak. I consider this really really ready now.

kim-em commented:
Nice! Thank @grunweg for making this happen, and @adomani  and @jcommelin for contributing documentation. I feel much better about our scripts/ directory now!

Next up, rewrite all the python ones in lean. :-)

bors merge

kim-em commented:
Haha, it caught a new one that has arrived on `master`!

kim-em commented:
bors merge

kim-em commented:
bors merge

grunweg commented:
> Next up, rewrite all the python ones in lean. :-)

@kim-em I do have some in-progress PRs regarding this; review welcome :-)
- #16334 and #16532 (depending on the former) rewrite two more style linters from Python in Lean
- #17555 rewrites a text-based linter as a syntax linter instead
- #17556 removes Python linters which have been rewritten as syntax linters
- #18275 and #18281 rewrite `init_creation.sh` as a text-based Lean linter
